_posts/2019-05-08-model-serving-in-pyorch.md

```diff
@@ -0,0 +1,58 @@
+
+### Model microservices
+
+If you're using your model in a server side context and you're managing multiple models, you might choose to treat each individual model (or each individual model version) as a separate service, usually using some sort of packaging mechanism like a Docker container. Then that service is often made network accessible via some sort of service, either using JSON over HTTP or an RPC technology like gRPC. The key characteristic of this approach is that you're defining a service with a single endpoint that just calls your model. Then you do do all of your model management (promotion, rollback, etc.) via whatever system you already use to manage your services (e.g. kubernetes, ECS).
+
+### Model servers
+
+An additional possible solution is to use a model server. This is an application built to manage and serve models. It allows you to upload multiple models and get distinct prediction endpoints for each of them. Typically such systems include a number of other features to help solve more of the whole problem of managing and serving models. This can include things like metrics, visualization, data pre-processing, and more. Even something as simple as having a system for automatically versioning models can make building important features like model rollbacks much easier.
+
+### Evolving Patterns
+
+The above is a somewhat arbitrary breakdown of different approaches based on a snapshot in time. Design patterns are still evolving. Recently, model server designs have started to adopt more of the technologies of general service infrastructure such as Docker containers and kubernetes, so many model servers have started to share properties of the model microservice design discussed above. For a deeper dive into the general concepts of model server designs, you can check out my [book on machine learning systems](https://www.manning.com/books/machine-learning-systems).
+
+## Serving PyTorch Models
+
+So, if you're a PyTorch user, what should you use if you want to take your models to production?
+
+If you're on mobile or working on an embedded system like a robot, direct embedding in your application is often the right choice. For mobile specifically, check out [this tutorial](https://pytorch.org/tutorials/advanced/super_resolution_with_caffe2.html) on deploying PyTorch models to mobile using ONNX. For other embedded systems like robots, running [inference on a PyTorch model from the C++ API](https://pytorch.org/tutorials/advanced/cpp_export.html) could be the right solution.
```

Comment:
I think we should be honest about this section that PyTorch -> ONNX is very limited. Or else, it'll be seen as a puff-piece, considering how much trouble our users already have with trying to export PyTorch -> ONNX (visible via bug reports and forum posts)

Reply:
Does this edit seem sufficiently moderate @soumith ?

Also, does this statement about our future mobile roadmap sound right, @jspisak ?

Reply:
I think this is the last outstanding question. Can you review and, if satisfied, merge, @soumith ?